File: backend/app/services/operation_log_service.ts
import OperationLog from '#models/operation_log'
import { DateTime } from 'luxon'
import db from '@adonisjs/lucid/services/db'
import { paginate, buildPaginationMeta } from '../utils/helpers.js'
import User from '#models/user'

interface LogIndexParams {
  page?: number
  perPage?: number
  userId?: number
  action?: string
  resourceType?: string
  startDate?: string
  endDate?: string
}

interface PaginatedResult<T> {
  data: T[]
  pagination: any
}

export default class OperationLogService {
  private static instance: OperationLogService

  static getInstance(): OperationLogService {
    if (!OperationLogService.instance) {
      OperationLogService.instance = new OperationLogService()
    }
    return OperationLogService.instance
  }

  static async log(
    userId: number | null,
    action: string,
    resourceType: string | null,
    resourceId?: number,
    oldValue?: Record<string, any>,
    newValue?: Record<string, any>,
    request?: {
      ip?: string
      url?: string
      method?: string
      userAgent?: string
    },
  ): Promise<OperationLog | null> {
    try {
      const log = new OperationLog()
      log.userId = userId
      log.action = action
      log.resourceType = resourceType
      log.resourceId = resourceId || null
      log.oldValues = oldValue || {}
      log.newValues = newValue || {}
      log.ipAddress = request?.ip || null
      log.userAgent = request?.userAgent || null
      await log.save()
      return log
    } catch (error) {
      console.error('Failed to write operation log:', error)
      return null
    }
  }

  async writeLog(
    userId: number | null,
    action: string,
    resourceType: string | null,
    resourceId?: number,
    oldValue?: Record<string, any>,
    newValue?: Record<string, any>,
    request?: {
      ip?: string
      url?: string
      method?: string
      userAgent?: string
    },
  ): Promise<OperationLog | null> {
    return OperationLogService.log(userId, action, resourceType, resourceId, oldValue, newValue, request)
  }

  async index(params: LogIndexParams): Promise<PaginatedResult<OperationLog>> {
    const { page, perPage, limit, offset } = paginate(params.page, params.perPage)

    const query = OperationLog.query()
      .preload('user', (q) => q.select('id', 'username', 'fullName', 'role'))
      .orderBy('createdAt', 'desc')

    if (params.userId) {
      query.where('userId', params.userId)
    }

    if (params.action) {
      query.whereILike('action', `%${params.action}%`)
    }

    if (params.resourceType) {
      query.where('resourceType', params.resourceType)
    }

    if (params.startDate) {
      query.where(
        'createdAt',
        '>=',
        DateTime.fromFormat(params.startDate, 'yyyy-MM-dd').startOf('day').toISO(),
      )
    }
    if (params.endDate) {
      query.where(
        'createdAt',
        '<=',
        DateTime.fromFormat(params.endDate, 'yyyy-MM-dd').endOf('day').toISO(),
      )
    }

    const cloneQuery = query.clone()
    const total = await cloneQuery.count('* as total')
    const totalCount = Number(total[0].$extras.total || 0)

    const data = await query.limit(limit).offset(offset).exec()

    return {
      data,
      pagination: buildPaginationMeta(totalCount, page, perPage),
    }
  }

  async getStats(days = 7): Promise<{
    total: number
    byAction: Record<string, number>
    byResource: Record<string, number>
    byUser: Array<{ userId: number; username: string; count: number }>
    daily: Array<{ date: string; count: number }>
  }> {
    const { start, end } = {
      start: DateTime.now().minus({ days: days - 1 }).startOf('day'),
      end: DateTime.now().endOf('day'),
    }

    const totalResult = await OperationLog.query()
      .where('createdAt', '>=', start.toISO())
      .where('createdAt', '<=', end.toISO())
      .count('* as total')

    const total = Number(totalResult[0].$extras.total || 0)

    const actionResults: any[] = await OperationLog.query()
      .where('createdAt', '>=', start.toISO())
      .where('createdAt', '<=', end.toISO())
      .select('action')
      .select(db.raw('COUNT(*) as action_count'))
      .groupBy('action')
      .orderByRaw('COUNT(*) DESC')
      .limit(10)
      .exec()

    const byAction: Record<string, number> = {}
    actionResults.forEach((r: any) => {
      byAction[r.action] = Number(r.$extras.action_count || 0)
    })

    const resourceResults: any[] = await OperationLog.query()
      .where('createdAt', '>=', start.toISO())
      .where('createdAt', '<=', end.toISO())
      .whereNotNull('resourceType')
      .select('resourceType')
      .select(db.raw('COUNT(*) as resource_count'))
      .groupBy('resourceType')
      .orderByRaw('COUNT(*) DESC')
      .limit(10)
      .exec()

    const byResource: Record<string, number> = {}
    resourceResults.forEach((r: any) => {
      byResource[r.resourceType!] = Number(r.$extras.resource_count || 0)
    })

    const userResults: any[] = await OperationLog.query()
      .where('operationLogs.createdAt', '>=', start.toISO())
      .where('operationLogs.createdAt', '<=', end.toISO())
      .whereNotNull('operationLogs.userId')
      .leftJoin('users', 'operationLogs.userId', 'users.id')
      .select('operationLogs.userId as user_id')
      .select(db.raw('COALESCE(users.username, ?) as username', ['Unknown']))
      .select(db.raw('COUNT(*) as user_count'))
      .groupBy('operationLogs.userId', 'users.username')
      .orderByRaw('COUNT(*) DESC')
      .limit(10)
      .exec()

    const byUser = userResults.map((r: any) => ({
      userId: Number(r.$extras.user_id),
      username: r.$extras.username,
      count: Number(r.$extras.user_count || 0),
    }))

    const dates: string[] = []
    let current = start
    while (current <= end) {
      dates.push(current.toFormat('yyyy-MM-dd'))
      current = current.plus({ days: 1 })
    }

    const dailyResults: any[] = await OperationLog.query()
      .where('createdAt', '>=', start.toISO())
      .where('createdAt', '<=', end.toISO())
      .select(db.raw('DATE(created_at) as date_key'))
      .select(db.raw('COUNT(*) as daily_count'))
      .groupByRaw('DATE(created_at)')
      .exec()

    const dailyMap: Record<string, number> = {}
    dailyResults.forEach((r: any) => {
      dailyMap[r.$extras.date_key] = Number(r.$extras.daily_count || 0)
    })

    const daily = dates.map((date) => ({
      date,
      count: dailyMap[date] || 0,
    }))

    return {
      total,
      byAction,
      byResource,
      byUser,
      daily,
    }
  }
}

File: backend/app/services/operation_log_service.py
from typing import Optional, Dict, Any
from datetime import datetime
from sqlalchemy.orm import Session
from sqlalchemy import and_, or_

from app.models.operation_log import OperationLog
from app.schemas.operation_log import OperationLogQuery
from app.schemas.common import PageResult


class OperationLogService:
    @staticmethod
    def get(db: Session, log_id: int) -> Optional[OperationLog]:
        return db.query(OperationLog).filter(OperationLog.id == log_id).first()

    @staticmethod
    def list(db: Session, query: OperationLogQuery) -> PageResult:
        q = db.query(OperationLog)

        if query.keyword:
            keyword = f"%{query.keyword}%"
            q = q.filter(or_(OperationLog.description.like(keyword), OperationLog.username.like(keyword)))

        if query.module:
            q = q.filter(OperationLog.module == query.module)

        if query.operation_type:
            q = q.filter(OperationLog.operation_type == query.operation_type)

        if query.user_id:
            q = q.filter(OperationLog.user_id == query.user_id)

        if query.status:
            q = q.filter(OperationLog.status == query.status)

        if query.start_time:
            q = q.filter(OperationLog.created_at >= query.start_time)

        if query.end_time:
            q = q.filter(OperationLog.created_at <= query.end_time)

        total = q.count()
        items = q.order_by(OperationLog.id.desc()).offset((query.page - 1) * query.page_size).limit(query.page_size).all()

        return PageResult(total=total, page=query.page, page_size=query.page_size, items=items)

    @staticmethod
    def create(
        db: Session,
        operation_type: str,
        module: str,
        description: str,
        user_id: Optional[int] = None,
        username: Optional[str] = None,
        ip_address: Optional[str] = None,
        user_agent: Optional[str] = None,
        request_method: Optional[str] = None,
        request_url: Optional[str] = None,
        request_params: Optional[Dict[str, Any]] = None,
        response_data: Optional[Dict[str, Any]] = None,
        old_data: Optional[Dict[str, Any]] = None,
        new_data: Optional[Dict[str, Any]] = None,
        status: str = "success",
        error_msg: Optional[str] = None,
        duration: Optional[int] = None,
    ) -> OperationLog:
        log = OperationLog(
            user_id=user_id,
            username=username,
            operation_type=operation_type,
            module=module,
            description=description,
            ip_address=ip_address,
            user_agent=user_agent,
            request_method=request_method,
            request_url=request_url,
            request_params=request_params,
            response_data=response_data,
            old_data=old_data,
            new_data=new_data,
            status=status,
            error_msg=error_msg,
            duration=duration,
        )
        db.add(log)
        db.commit()
        db.refresh(log)
        return log
